docs/orchestrating-elastic-stack-applications/elasticsearch/users-and-roles.asciidoc

```diff
@@ -0,0 +1,154 @@
+
+If you specify multiple users with the same name in more than one secret, the last one takes precedence.
+If you specify multiple roles with the same name in more than one secret, a single entry per role is be derived from the concatenation of its corresponding users from all secrets.
+
+The following Secret specifies three users and their respective roles:
+
+[source,yaml]
+----
+kind: Secret
+apiVersion: v1
+metadata:
+  name: my-filerealm-secret
+stringData:
+  users: |-
+    rdeniro:$2a$10$BBJ/ILiyJ1eBTYoRKxkqbuDEdYECplvxnqQ47uiowE7yGqvCEgj9W
+    alpacino:$2a$10$cNwHnElYiMYZ/T3K4PvzGeJ1KbpXZp2PfoQD.gfaVdImnHOwIuBKS
+    jacknich:{PBKDF2}50000$z1CLJt0MEFjkIK5iEfgvfnA6xq7lF25uasspsTKSo5Q=$XxCVLbaKDimOdyWgLCLJiyoiWpA/XDMe/xtVgn1r5Sg=
+  users_roles: |-
+    admin:rdeniro
+    power_user:alpacino,jacknich
+    user:jacknich
+----
+
+You can populate the content of both `users` and `users_roles` using the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/users-command.html[elasticsearch-users] tool.
+
+For example, invoking the tool in a Docker container:
+
+[source,sh]
+----
+# create a folder with the 2 files
+mkdir filerealm
+touch filerealm/users filerealm/users_roles
+
+# create user 'myuser' with role 'monitoring_user'
+docker run \
+    -v ${PWD}/filerealm:/usr/share/elasticsearch/config \
+    docker.elastic.co/elasticsearch/elasticsearch:{version} \
+    bin/elasticsearch-users useradd myuser -p mypassword -r monitoring_user
+
+# create a Kubernetes secret with the file realm content
+kubectl create secret generic my-file-realm-secret --from-file filerealm
+----
+
+== Creating custom roles
+
+link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html[Roles] can be specified using the
+link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html#roles-management-api[Role management API],
+or the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html#roles-management-ui[Role management UI in Kibana].
+
+Additionally, link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html#roles-management-file[file-based role management] can be achieved by referencing Kubernetes secrets containing the roles specification.
+
+[source,yaml]
+----
+apiVersion: elasticsearch.k8s.elastic.co/{eck_crd_version}
+kind: Elasticsearch
+metadata:
+  name: elasticsearch-sample
+spec:
+  version: {version}
+  auth:
+    roles:
+    - secretName: my-roles-secret-1
+    - secretName: my-roles-secret-2
+  nodeSets:
+  - name: default
+    count: 1
+----
+
+Several secrets can be referenced in the Elasticsearch specification.
+ECK aggregates their content into a single secret, mounted in every Elasticsearch Pod.
+
+Each secret must have a `roles.yml` entry, containing the link:https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html#roles-management-file[roles configuration].
```

Comment:
"roles configuration" sounds wrong (because it is the configuration of the roles) but I'm not sure how it could be improved. "roles' configuration" also doesn't seem quite right. Plural "roles" is correct because you can define multiple roles in the config. I'm fine with it as is but wanted to raise it in case someone had a better idea.

Reply:
"roles specification" instead of "roles configuration"?

Reply:
Or maybe "role definition" since the ES doc is called "defining roles".

Reply:
+1 for "roles definition"